File: frontend/src/utils/htmlUtils.test.ts
import { describe, it, expect } from 'vitest';
import { hasTextContent, stripHtmlTags } from './htmlUtils';

describe('htmlUtils', () => {
  describe('hasTextContent', () => {
    // ========== REGRESSION: Empty Quill HTML Validation ==========
    // Bug: Quill's empty editor produces <p><br></p> which passes .trim() check
    // but should be treated as empty content

    it('REGRESSION: should reject Quill empty editor markup', () => {
      // Quill's default empty state
      expect(hasTextContent('<p><br></p>')).toBe(false);
      expect(hasTextContent('<p><br/></p>')).toBe(false);
      expect(hasTextContent('<p><br /></p>')).toBe(false);
    });

    it('REGRESSION: should reject whitespace-only HTML', () => {
      expect(hasTextContent('<p>   </p>')).toBe(false);
      expect(hasTextContent('<p>\n\t </p>')).toBe(false);
      expect(hasTextContent('<div>  <span>  </span>  </div>')).toBe(false);
    });

    it('REGRESSION: should reject empty nested tags', () => {
      expect(hasTextContent('<p><strong></strong></p>')).toBe(false);
      expect(hasTextContent('<p><em><strong></strong></em></p>')).toBe(false);
      expect(hasTextContent('<div><p></p></div>')).toBe(false);
    });

    // ========== Valid Content Tests ==========

    it('should accept HTML with actual text content', () => {
      expect(hasTextContent('<p>Hello</p>')).toBe(true);
      expect(hasTextContent('<p>Hello <strong>World</strong></p>')).toBe(true);
      expect(hasTextContent('<div><p>Text</p></div>')).toBe(true);
    });

    it('should accept plain text without HTML', () => {
      expect(hasTextContent('Plain text')).toBe(true);
      expect(hasTextContent('Text with spaces')).toBe(true);
    });

    it('should handle special characters', () => {
      expect(hasTextContent('<p>!@#$%</p>')).toBe(true);
      expect(hasTextContent('<p>123</p>')).toBe(true);
      expect(hasTextContent('<p>&nbsp;</p>')).toBe(false); // Non-breaking space is still whitespace
    });

    // ========== Edge Cases ==========

    it('should handle null and undefined', () => {
      expect(hasTextContent(null)).toBe(false);
      expect(hasTextContent(undefined)).toBe(false);
      expect(hasTextContent('')).toBe(false);
    });

    it('should handle malformed HTML', () => {
      expect(hasTextContent('<p>Unclosed tag')).toBe(true);
      expect(hasTextContent('No tags at all')).toBe(true);
      // Note: <><> behavior with DOM parsing may vary
    });

    // ========== REGRESSION: Frontend/Backend Consistency ==========
    // Bug: Frontend used trim() while backend used isBlank()
    // This caused confusing validation errors where frontend passed but backend rejected

    it('REGRESSION: should reject non-breaking spaces only (nbsp)', () => {
      // Multiple nbsp characters - common in rich text editors
      expect(hasTextContent('<p>&nbsp;</p>')).toBe(false);
      expect(hasTextContent('<p>&nbsp;&nbsp;&nbsp;</p>')).toBe(false);
      expect(hasTextContent('<div>&nbsp;<span>&nbsp;</span></div>')).toBe(false);
    });

    it('REGRESSION: should reject zero-width spaces', () => {
      // Zero-width space (U+200B) - invisible character
      const zws = '\u200B';
      expect(hasTextContent(`<p>${zws}</p>`)).toBe(false);
      expect(hasTextContent(`<p>${zws}${zws}${zws}</p>`)).toBe(false);
      expect(hasTextContent(`<p> ${zws} </p>`)).toBe(false);
    });

    it('REGRESSION: should reject other Unicode whitespace', () => {
      // Thin space (U+2009), hair space (U+200A), ideographic space (U+3000)
      expect(hasTextContent('<p>\u2009</p>')).toBe(false);
      expect(hasTextContent('<p>\u200A</p>')).toBe(false);
      expect(hasTextContent('<p>\u3000</p>')).toBe(false);

      // Mixed Unicode whitespace
      expect(hasTextContent('<p>&nbsp;\u2009\u200A</p>')).toBe(false);
    });

    it('REGRESSION: should accept text with Unicode whitespace inside', () => {
      // Valid text that contains Unicode whitespace is still valid
      expect(hasTextContent('<p>Hello&nbsp;World</p>')).toBe(true);
      expect(hasTextContent('<p>Valid\u200BText</p>')).toBe(true);
      expect(hasTextContent('<p>Japanese\u3000Space</p>')).toBe(true);
    });
  });

  describe('stripHtmlTags', () => {
    it('should remove HTML tags and return plain text', () => {
      expect(stripHtmlTags('<p>Hello</p>')).toBe('Hello');
      expect(stripHtmlTags('<p>Hello <strong>World</strong></p>')).toBe('Hello World');
      expect(stripHtmlTags('<div><p>Nested</p></div>')).toBe('Nested');
    });

    it('should handle empty content', () => {
      expect(stripHtmlTags('')).toBe('');
      expect(stripHtmlTags(null)).toBe('');
      expect(stripHtmlTags(undefined)).toBe('');
      expect(stripHtmlTags('<p></p>')).toBe('');
    });

    it('should trim whitespace', () => {
      expect(stripHtmlTags('<p>  Text  </p>')).toBe('Text');
      expect(stripHtmlTags('  <p>Text</p>  ')).toBe('Text');
    });
  });
});

File: src/test/java/io/indcloud/util/HtmlUtilsTest.java
package io.indcloud.util;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Tests for HtmlUtils - HTML content validation and processing.
 * Focus on regression testing for empty Quill editor markup validation.
 */
class HtmlUtilsTest {

    // ========== REGRESSION: Empty Quill HTML Validation ==========
    // Bug: Quill's empty editor produces <p><br></p> which passes .trim() check
    // but should be treated as empty content

    @Test
    void hasTextContent_shouldRejectQuillEmptyEditorMarkup() {
        // Quill's default empty state variations
        assertThat(HtmlUtils.hasTextContent("<p><br></p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p><br/></p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p><br /></p>")).isFalse();
    }

    @Test
    void hasTextContent_shouldRejectWhitespaceOnlyHtml() {
        assertThat(HtmlUtils.hasTextContent("<p>   </p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p>\n\t </p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<div>  <span>  </span>  </div>")).isFalse();
    }

    @Test
    void hasTextContent_shouldRejectEmptyNestedTags() {
        assertThat(HtmlUtils.hasTextContent("<p><strong></strong></p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p><em><strong></strong></em></p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<div><p></p></div>")).isFalse();
    }

    // ========== REGRESSION: Unicode Whitespace Tests ==========
    // Bug: trim() only strips ASCII whitespace ≤0x20, not Unicode whitespace
    // Payloads like <p>&nbsp;</p> would pass validation and persist

    @Test
    void REGRESSION_hasTextContent_shouldRejectNonBreakingSpaces() {
        // &nbsp; entity (U+00A0) - common in Quill/rich text editors
        assertThat(HtmlUtils.hasTextContent("<p>&nbsp;</p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p>&nbsp;&nbsp;&nbsp;</p>")).isFalse();

        // Mixed with ASCII whitespace
        assertThat(HtmlUtils.hasTextContent("<p> &nbsp; </p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<div>&nbsp;<span>&nbsp;</span>&nbsp;</div>")).isFalse();
    }

    @Test
    void REGRESSION_hasTextContent_shouldRejectZeroWidthSpaces() {
        // Zero-width space (U+200B)
        String zeroWidthSpace = "\u200B";
        assertThat(HtmlUtils.hasTextContent("<p>" + zeroWidthSpace + "</p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p>" + zeroWidthSpace + zeroWidthSpace + "</p>")).isFalse();

        // Mixed with other whitespace
        assertThat(HtmlUtils.hasTextContent("<p> " + zeroWidthSpace + " </p>")).isFalse();
    }

    @Test
    void REGRESSION_hasTextContent_shouldRejectOtherUnicodeWhitespace() {
        // Various Unicode whitespace characters
        String thinSpace = "\u2009";      // Thin space (U+2009)
        String hairSpace = "\u200A";      // Hair space (U+200A)
        String ideographicSpace = "\u3000"; // Ideographic space (U+3000)

        assertThat(HtmlUtils.hasTextContent("<p>" + thinSpace + "</p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p>" + hairSpace + "</p>")).isFalse();
        assertThat(HtmlUtils.hasTextContent("<p>" + ideographicSpace + "</p>")).isFalse();

        // Mixed Unicode whitespace
        assertThat(HtmlUtils.hasTextContent("<p>&nbsp;" + thinSpace + hairSpace + "</p>")).isFalse();
    }

    @Test
    void REGRESSION_hasTextContent_shouldAcceptTextWithUnicodeWhitespace() {
        // Valid text that happens to contain Unicode whitespace
        assertThat(HtmlUtils.hasTextContent("<p>Hello&nbsp;World</p>")).isTrue();
        assertThat(HtmlUtils.hasTextContent("<p>Valid\u200BText</p>")).isTrue();
        assertThat(HtmlUtils.hasTextContent("<p>Japanese\u3000Space</p>")).isTrue();
    }

    // ========== Valid Content Tests ==========

    @Test
    void hasTextContent_shouldAcceptHtmlWithActualText() {
        assertThat(HtmlUtils.hasTextContent("<p>Hello</p>")).isTrue();
        assertThat(HtmlUtils.hasTextContent("<p>Hello <strong>World</strong></p>")).isTrue();
        assertThat(HtmlUtils.hasTextContent("<div><p>Text</p></div>")).isTrue();
    }

    @Test
    void hasTextContent_shouldAcceptPlainTextWithoutHtml() {
        assertThat(HtmlUtils.hasTextContent("Plain text")).isTrue();
        assertThat(HtmlUtils.hasTextContent("Text with spaces")).isTrue();
    }

    @Test
    void hasTextContent_shouldHandleSpecialCharacters() {
        assertThat(HtmlUtils.hasTextContent("<p>!@#$%</p>")).isTrue();
        assertThat(HtmlUtils.hasTextContent("<p>123</p>")).isTrue();
    }

    // ========== Edge Cases ==========

    @Test
    void hasTextContent_shouldHandleNullAndEmpty() {
        assertThat(HtmlUtils.hasTextContent(null)).isFalse();
        assertThat(HtmlUtils.hasTextContent("")).isFalse();
    }

    @Test
    void hasTextContent_shouldHandleMalformedHtml() {
        assertThat(HtmlUtils.hasTextContent("<p>Unclosed tag")).isTrue();
        assertThat(HtmlUtils.hasTextContent("No tags at all")).isTrue();
        // Note: <><> is parsed by Jsoup as empty tags, behavior may vary
    }

    // ========== stripHtmlTags Tests ==========

    @Test
    void stripHtmlTags_shouldRemoveTagsAndReturnPlainText() {
        assertThat(HtmlUtils.stripHtmlTags("<p>Hello</p>")).isEqualTo("Hello");
        assertThat(HtmlUtils.stripHtmlTags("<p>Hello <strong>World</strong></p>")).isEqualTo("Hello World");
        assertThat(HtmlUtils.stripHtmlTags("<div><p>Nested</p></div>")).isEqualTo("Nested");
    }

    @Test
    void stripHtmlTags_shouldHandleEmptyContent() {
        assertThat(HtmlUtils.stripHtmlTags("")).isEmpty();
        assertThat(HtmlUtils.stripHtmlTags(null)).isEmpty();
        assertThat(HtmlUtils.stripHtmlTags("<p></p>")).isEmpty();
    }

    @Test
    void stripHtmlTags_shouldTrimWhitespace() {
        assertThat(HtmlUtils.stripHtmlTags("<p>  Text  </p>")).isEqualTo("Text");
        assertThat(HtmlUtils.stripHtmlTags("  <p>Text</p>  ")).isEqualTo("Text");
    }
}
